client/app/pages/queries/utils/useVisualizationTabHandler.js

```diff
@@ -0,0 +1,30 @@
+import { useState, useEffect, useMemo } from "react";
+import { first, orderBy, isFunction } from "lodash";
+import { $location, $rootScope } from "@/services/ng";
+
+function updateUrlHash(...args) {
+  $location.hash(...args);
+  $rootScope.$applyAsync();
+}
+
+export default function useVisualizationTabHandler(visualizations, onAddVisualizationInUrl) {
+  const firstVisualization = useMemo(() => first(orderBy(visualizations, ["id"])), [visualizations]);
+  const [selectedTab, setSelectedTab] = useState(+$location.hash() || firstVisualization.id);
+
+  useEffect(() => {
+    if ($location.hash() !== "add") {
+      updateUrlHash(selectedTab !== firstVisualization.id ? selectedTab : null);
+    }
+  }, [firstVisualization.id, selectedTab]);
+
+  useEffect(() => {
+    if ($location.hash() === "add") {
```

Comment:
This is a legacy which we should avoid in React code:

https://github.com/getredash/redash/blob/bc52b788892181f50d419048ce2879148feae6e7/client/app/pages/queries/view.js#L465-L470

Let's just drop it and I'll find another solution later

Reply:
I thought just the part below of it was the "TODO" part :sweat_smile:, my bad. Anyway, this makes my life easier as that `onAddInUrl` needed a different function in the View :grin:.

Reply:
The only purpose of this logic was to open Add Visualization dialog for new queries: query is saved, url changes, Angular destroys old controller and creates new one - so it's the way to signal new controller that it should open Add Visualization dialog. This flow is related only to Query Source controller and it's impossible with Query View page (because it cannot open unsaved query)